test/multiply_test.cc

```diff
@@ -244,24 +307,37 @@ template <class Integer> void SlowRefInt(const Integer *A, const Integer *B, flo
 }
 
 
-void Compare(const float *float_ref, const float *int_ref, const float *int_test, std::size_t size) {
+void Compare(const float *float_ref, const float *int_ref, const float *int_test, std::size_t size, std::string test_info,
+ float int_tolerance, float float_tolerance, float MSE_float_tolerance, float MSE_int_tolerance) {
   float int_sum = 0.0, float_sum = 0.0;
   for (std::size_t i = 0; i < size; ++i) {
     float int_diff = int_ref[i] - int_test[i];
     float float_diff = float_ref[i] - int_test[i];
-/*    if (fabs(int_diff) > .1 || fabs(float_diff) > 1) {
-      std::cerr << "Inaccurate at " << i << ' ' << float_ref[i] << ' ' << int_ref[i] << ' ' << int_test[i] << '\n';
-    }*/
+    CHECK_MESSAGE(fabs(int_diff) <= int_tolerance, test_info << "Inaccurate compared to int reference at " << i << ' ' << int_ref[i] << ' ' << int_test[i]);
+    CHECK_MESSAGE(fabs(float_diff) <= float_tolerance, test_info << "Inaccurate compared to float reference at " << i << ' ' << float_ref[i] << ' ' << int_test[i]);
     int_sum += int_diff * int_diff;
     float_sum += float_diff * float_diff;
   }
-  std::cout << "Float MSE = " << sqrt(float_sum / size) << "\tInt MSE = " << sqrt(int_sum / size) << std::endl;
+  CHECK_MESSAGE(fabs(sqrt(float_sum / size)) <= MSE_float_tolerance, test_info << "Float MSE = " << sqrt(float_sum / size));
+  CHECK_MESSAGE(fabs(sqrt(int_sum / size)) <= MSE_int_tolerance, test_info << "Int MSE = " << sqrt(int_sum / size));
+  //We want to see a test failure when we fix some of the accumulation errors, so that we also
+  //fix our tests after we fix the issue. The easiest way to test this is through the MSE
+  if (MSE_float_tolerance > 0.1) {
+    CHECK_FALSE_MESSAGE(fabs(sqrt(float_sum / size)) <= MSE_float_tolerance/2, test_info << "Float MSE = " << sqrt(float_sum / size) <<
+      " much better than the expected: " << MSE_float_tolerance);
+  }
+
+  if (MSE_int_tolerance > 0.03) {
+    CHECK_FALSE_MESSAGE(fabs(sqrt(int_sum / size)) <= MSE_int_tolerance/2, test_info << "Int MSE = " << sqrt(int_sum / size) << 
+      " much better than the expected: " << MSE_int_tolerance);
+  }
 }
 
-template <class Routine> void TestMultiply(Index A_rows, Index width, Index B_cols) {
+template <class Routine> void TestMultiply(Index A_rows, Index width, Index B_cols,
```

Comment:
At some point we start developing a struct to pass tolerances, but I see why you did this for now